### DISCUSSION TITLE
Version pinning in Universe

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#1873] [General]
## Overview

Dagger Universe is a catalog of CUE packages, carefully curated for safety and quality, that all Dagger developers can use to build awesome CICD pipelines without reinventing the wheel.

It is crucial to the success of Dagger that Universe can combine four properties:
1) Growth. We need a *lot* of packages, and the capacity to publish more and more at an accelerated growth rate
2) Quality. Packages must work well, be well documented, bugs need to be fixed quickly, etc.
3) Safety. Using Dagger Universe packages must always make users more safe, and never less safe!
3) Freshness: always remain up-to-date with ustream releases.

This is a (very) difficult thing to achieve, but the effort is well worth it. For comparison think of Linux distributions like Red Hat or Debian. This requires a similar level of investment - but in exchange it can yield similar levels of ubiquity and platform power.

One dimension of this tradeoff is *version pinning*. Should universe packages pin their dependencies to upstream software, or not?

* To answer “yes” means higher quality and safety, but considerably more maintenance work. If we fail to scale the maintenance work, we can hurt freshness, growth, or both.
* To answer “no” means less maintenance work, and one less obstacle to freshness and growth; but at the cost of lower quality and safety.

Let’s discuss this important treadeoff here :)

## Tentative decision

Based on the discussion below, the consensus seems to be:

* *Yes*, Universe packages should pin their own dependencies
* There are several mechanisms for doing this. Which one we should use is to be determined.
* We should make an incremental step towards this goal as part of the 0.2 Europa release; but it's not yet determined what work should be done now versus later.
* Next step is for 0.2 release owner (@aluzzardi) to determine work needed now, and track it in another issue

**Comment by talentedmrjones** [comment]
We could automate updating versions but are we ok with that script committing back to the repo?

**Comment by aluzzardi** [comment]
I was discussing this earlier with @samalba: there is an overkill workaround.

If we were to use #Dockerfile with embedding, we’d be kept up to date by dependabot (automatically opening PRs against Dockerfiles when upstream gets updated).

Same principle applies to other package managers, e.g. if instead of doing a yarn install of Netlify at a specific version we embed a package.json, then the dependency gets updated automatically.

This applies to all kind of things (e.g. pip install aws-cli vs embedding a requirements.txt that gets auto-updated).

So either we externalise as much as possible and let the existing language tool chains deal with updates, or we implement our own tool chain to update CUE references to external packages

**Comment by Alvise88** [comment]
I prefer option 1, couldn't cuelang be used to bind supported versions ?
Dagger/Cue could register a package behaving like https://github.com/hashicorp/go-version calling this package semver.

I guess it is possible to do something like this:

```cue
import semver

#Kubectl: {
  version: string & semver.NewConstraint(">= 1.19, < 1.24")
}
```

I don't know if it's feasible

**Comment by shykes** [comment]
> I prefer option 1, couldn't cuelang be used to bind supported versions ?

To clarify, what are you referring to by “option 1”?

**Comment by Alvise88** [comment]
> > I prefer option 1, couldn't cuelang be used to bind supported versions ?
> 
> To clarify, what are you referring to by “option 1”?

Option 1: To answer “yes” means higher quality and safety, but considerably more maintenance work. If we fail to scale the maintenance work, we can hurt freshness, growth, or both.

I think universe packages have to pin their dependencies to upstream software

**Comment by shykes** [comment]
>> Option 1: To answer “yes” means higher quality and safety, but considerably more maintenance work. If we fail to scale the maintenance work, we can hurt freshness, growth, or both.

> I think universe packages have to pin their dependencies to upstream software

OK, thanks for clarifying! I agree it’s the most attractive option. But the additional work is not negligeable, so it requires careful planning and excellent execution to pull off.

**Comment by jlongtine** [comment]
Another thing that occurred to me reading this is `universe` <-> `engine` API  versions. We _hopefully_ won't have a need to update `engine` APIs very often, but it seems pretty likely it will happen at least occasionally. So we should probably keep that in mind for the future, at the very least. More of a 1.0 thing, likely.

**Comment by shykes** [comment]
This is an open discussion topic, rather than a clear call to action. @aluzzardi in your opinion, what actions should we take now (in time for 0.2-beta.1) versus later? I suggest opening a separate issue for that, and moving this one out of the milestone.

**Comment by aluzzardi** [comment]
- **Now**: Universe packages should pin their own dependencies.
  - Seems to be the overall agreement in here, and it makes sense to me (tradeoff seems good to me)
- **Now**: Manual enforcing. I think we have our bases covered through `alpine` (base of most other packages). We should keep an eye out when merging new packages.
- **Later**: Automatic enforcing. Though to do for **everything**, but at least we could start with the most common cases (e.g. force `{docker/engine}.#Pull` to have a digest, possibly as a "linter" step )
- **Later**: Keep it fresh. Same as above, we could start with automation around the most common ones (e.g. docker images), gradually work our way up

The only unknown for me is whether we want to do this specifically for universe or for *everything* (e.g. 3rd party actions as well).

If it's only for universe, then the above plan is enough. However, if we want to do this for everything then we should change `engine.#Pull` to require a digest, to avoid a breaking change in the future.

**Comment by shykes** [comment]
I'm converting this to a discussion, since there is no clear path to closing it.

**Comment by mabis** [comment]
late to the party, I tried to search around before commenting but haven't managed to find a confirmation one why or the other so: 

since safety is an explict, important property shouldn't dagger encourage pinning to sha rather than tags? (git) tags are mutable and what commit they point to can be changed and if that's done by malicious actors we all know the results

**Reply by shykes** [reply]
Yes, absolutely. When I talk about version pinning I mean pinning by digest, for the reasons you mentioned.

**Reply by mabis** [reply]
great, the thought occurred to me while going through the docs and seeing:

`# To pin external dependencies, you can use 'project update github.com/[package-source]@v[n]'`

in the integrating with github actions example
